include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -118,6 +119,27 @@ struct getOrientedFresnel<F, true NBL_PARTIAL_REQ_BOT(fresnel::TwoSidedFresnel<F
         return fresnel.getReorientedFresnel(NdotV);
     }
 };
+
+template<class N, class LS, class Interaction, class MicrofacetCache>
+struct overwrite_DG
+{
+    using scalar_type = typename N::scalar_type;
+    using quant_type = typename N::quant_type;
+    using quant_query_type = typename N::quant_query_type;
+    using g2g1_query_type = typename N::g2g1_query_type;
+    NBL_CONSTEXPR_STATIC_INLINE bool HasOverwrite = ndf::NDF_CanOverwriteDG<N>;
+
+    template<typename C=bool_constant<!HasOverwrite> >
+    static enable_if_t<C::value && !HasOverwrite, void> __call(NBL_REF_ARG(scalar_type) DG, N ndf, NBL_CONST_REF_ARG(g2g1_query_type) query, NBL_CONST_REF_ARG(quant_query_type) quant_query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+    }
+    template<typename C=bool_constant<HasOverwrite> >
+    static enable_if_t<C::value && HasOverwrite, void> __call(NBL_REF_ARG(scalar_type) DG, N ndf, NBL_CONST_REF_ARG(g2g1_query_type) query, NBL_CONST_REF_ARG(quant_query_type) quant_query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        quant_type dg = ndf.template Dcorrelated<LS, Interaction, MicrofacetCache>(query, quant_query, _sample, interaction, cache);
+        DG = dg.projectedLightMeasure;
+    }
```

Comment:
cool, but if you're gonna do it this way, you could even stick the methods in `SCookTorrance`  which would save you a lot of typedef aliases

which kinda gives me an idea that all the `impl` functors could be done and hidden this way

Reply:
even `getOrientedFresnel` and `checkValid`